File: app/src/main/java/com/github/aakumykov/sync_dir_to_cloud/domain/use_cases/sync_task/SyncTaskManagingUseCase.java
package com.github.aakumykov.sync_dir_to_cloud.domain.use_cases.sync_task;

import androidx.lifecycle.LiveData;

import com.github.aakumykov.sync_dir_to_cloud.domain.entities.SyncTask;
import com.github.aakumykov.sync_dir_to_cloud.interfaces.iSyncTaskManager;
import com.github.aakumykov.sync_dir_to_cloud.interfaces.iSyncTaskUpdater;

import java.util.List;

public class SyncTaskManagingUseCase {

    private final iSyncTaskManager mSyncTaskManager;
    private final iSyncTaskUpdater mISyncTaskUpdater;

    public SyncTaskManagingUseCase(iSyncTaskManager syncTaskManager,
                                   iSyncTaskUpdater syncTaskUpdater) {
        mSyncTaskManager = syncTaskManager;
        mISyncTaskUpdater = syncTaskUpdater;
    }


    public LiveData<List<SyncTask>> listSyncTasks() {
        return mSyncTaskManager.listSyncTasks();
    }


    public void addSyncTask(final SyncTask syncTask) {
        mSyncTaskManager.createSyncTask(syncTask);
    }


    public void updateSyncTask(final SyncTask syncTask) {
        mISyncTaskUpdater.updateSyncTask(syncTask);
    }


    public SyncTask getSyncTask(final String id) {
        return mSyncTaskManager.getSyncTask(id);
    }


    public void deleteSyncTask(final SyncTask syncTask) {
        mSyncTaskManager.deleteSyncTask(syncTask);
    }
}

File: app/src/main/java/com/github/aakumykov/sync_dir_to_cloud/domain/use_cases/sync_task/SyncTaskManagingUseCase.kt
package com.github.aakumykov.sync_dir_to_cloud.domain.use_cases.sync_task

import androidx.lifecycle.LiveData
import com.github.aakumykov.sync_dir_to_cloud.di.annotations.AppScope
import com.github.aakumykov.sync_dir_to_cloud.domain.entities.SyncTask
import com.github.aakumykov.sync_dir_to_cloud.interfaces.for_repository.sync_task.SyncTaskCreatorDeleter
import com.github.aakumykov.sync_dir_to_cloud.interfaces.for_repository.sync_task.SyncTaskReader
import com.github.aakumykov.sync_dir_to_cloud.interfaces.for_repository.sync_task.SyncTaskUpdater
import javax.inject.Inject

@AppScope
class SyncTaskManagingUseCase @Inject constructor(
    val syncTaskReader: SyncTaskReader,
    private val syncTaskCreatorDeleter: SyncTaskCreatorDeleter,
    private val syncTaskUpdater: SyncTaskUpdater
) {

    suspend fun listSyncTasks(): LiveData<List<SyncTask>> {
        return syncTaskReader.listSyncTasks()
    }


    suspend fun getSyncTask(id: String): SyncTask {
        return syncTaskReader.getSyncTask(id)
    }


    suspend fun deleteSyncTask(syncTask: SyncTask) {
        syncTaskCreatorDeleter.deleteSyncTask(syncTask)
    }

    suspend fun deleteSyncTask(taskId: String) {
        syncTaskCreatorDeleter.deleteSyncTask(taskId)
    }


    suspend fun createOrUpdateSyncTask(syncTask: SyncTask) {
        if (null != syncTaskReader.getSyncTask(syncTask.id))
            syncTaskUpdater.updateSyncTask(syncTask)
        else
            syncTaskCreatorDeleter.createSyncTask(syncTask)
    }
}